File: src/test/java/leetcode/LC216CombinationSumIII.java
package leetcode;

import org.junit.Test;

import java.util.*;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/combination-sum-iii/
 */
public class LC216CombinationSumIII {
    public class Solution {
        private void traverse(int contains, int sum, int left, int n, List<Integer> found, int start) {
            if (left == 0 && sum == n) {
                found.add(contains);
            } else {
                for (int i = start; i <= 9; i++) {
                    int mask = 1 << i;
                    if ((contains & mask) == 0) {
                        traverse(contains | mask, sum + i, left - 1, n, found, i + 1);
                    }
                }
            }
        }

        public List<List<Integer>> combinationSum3(int k, int n) {
            List<Integer> found = new ArrayList<>();
            traverse(0, 0, k, n, found, 1);
            List<List<Integer>> all = new ArrayList<>();
            for (int contains : found) {
                List<Integer> entry = new ArrayList<>();
                all.add(entry);
                for (int i = 1; i <= 9; i++) {
                    int mask = 1 << i;
                    if ((contains & mask) != 0) {
                        entry.add(i);
                    }
                }
            }
            return all;
        }
    }

    private void test(int[][] expected, List<List<Integer>> result) {
        for (List<Integer> entry : result) {
            Collections.sort(entry);
        }
        //Collections.sort(result, );
        assertEquals(expected.length, result.size());
        for (int i = 0; i < expected.length; i++) {
            List<Integer> entry = result.get(i);
            assertEquals(expected[i].length, entry.size());
            for (int j = 0; j < expected[i].length; j++) {
                assertEquals(expected[i][j], entry.get(j).intValue());
            }
        }
    }

    @Test
    public void test_3_7() throws Exception {
        int[][] expected = {{1, 2, 4}};
        test(expected, new Solution().combinationSum3(3, 7));
    }

    @Test
    public void test_3_9() throws Exception {
        int[][] expected = {{1, 2, 6}, {1, 3, 5}, {2, 3, 4}};
        test(expected, new Solution().combinationSum3(3, 9));
    }
}

File: algorithms/code/leetcode/lc216_combination_sum_iii/lc216_combination_sum_iii.py
# https://leetcode.com/problems/combination-sum-iii/

import unittest


class Solution:
    # pylint: disable=R0913,R0917
    def __traverse(
        self,
        contains: int,
        summed: int,
        left: int,
        n: int,
        found: list[int],
        start: int,
    ) -> None:
        if left == 0 and summed == n:
            found.append(contains)
        else:
            for i in range(start, 10):
                mask = 1 << i
                # if (contains & mask) == 0:
                self.__traverse(
                    contains | mask, summed + i, left - 1, n, found, i + 1
                )

    def combinationSum3(self, k: int, n: int) -> list[list[int]]:
        found: list[int] = []
        self.__traverse(0, 0, k, n, found, 1)
        each = []
        for contains in found:
            entry: list[int] = []
            each.append(entry)
            for i in range(1, 10):
                mask = 1 << i
                if (contains & mask) != 0:
                    entry.append(i)
        return each


class TestCode(unittest.TestCase):
    def __test(
        self, expected: list[list[int]], result: list[list[int]]
    ) -> None:
        for entry in result:
            entry.sort()
        assert len(expected) == len(result)
        for i, expected_i in enumerate(expected):
            entry = result[i]
            assert len(expected_i) == len(entry)
            for j, expected_i_j in enumerate(expected_i):
                assert expected_i_j == entry[j]

    def test_3_7(self) -> None:
        expected = [[1, 2, 4]]
        self.__test(expected, Solution().combinationSum3(3, 7))

    def test_3_9(self) -> None:
        expected = [[1, 2, 6], [1, 3, 5], [2, 3, 4]]
        self.__test(expected, Solution().combinationSum3(3, 9))
